### ISSUE TITLE
sfizz.lv2 breaking the Ardour5 "Controllers" automation?

### ISSUE THREAD

**Issue by lminiero** [sfztools/sfizz#325]
Hi all,

I'm getting a weird issue when using sfizz as a plugin in Ardour5. Basically, for tracks that equip the sfizz lv2 plugin, I can't configure any controller-related automation: when going over the "Controllers" menu in the Automation list, nothing appears apart from a thin line, that makes me suspect Ardour5 is actually trying to show an empty list. This doesn't seem to happen for other MIDI tracks using different plugins (e.g., Calf Fluidsynth), where the proper sublist appears ("Controllers 0-15", "Controllers 16-31", etc.).

I think this started happening in a recent update, as I most definitely used Controller automation in the past with Ardour5/sfizz, since it was a key component of an effort I invested on using CC messages to perform switches in .sfz files. I just pulled the latest version of the sfizz `develop` branch to make sure it wasn't something that had been fixed already, but that didn't seem to help. Not sure if this is because of some change in Ardour5 or sfizz, but as I said this only seems to happen with sfizz-powered tracks, so I thought I'd open an issue here first to see if this rang any bell.

Thanks for your amazing work!

**Comment by paulfd** [comment]
I think this is related to MIDNAM, I've seen it but never investigated well because, well, I don't really know how it's supposed to work in Ardour :D

I'll look into it!

**Comment by paulfd** [comment]
Right now by default the MIDNAM outputs an "empty" list of automation unless you use `label_cc` opcodes to name the controllers.
I've added some logic in #326 to also output `Unnamed` controllers in the MIDNAM. All the controllers used by the SFZ file that sfizz knows how to handle should then be in Ardour's list, or so it should in version 6. I'll try to install version 5 to test, or you can try and build the pull request from source !

**Comment by lminiero** [comment]
Wow, thanks for the super-quick answer! I'm away from my laptop, at the moment, but I should be able to compile and test in a couple of hours, so I'll make sure to give you some feedback as soon as possible.

**Comment by lminiero** [comment]
Something changed, now, but I'm not sure it's fixed. If I list controllers, I get `Controllers 0-0`, which has as its only item `1: Unnamed [1]`. If I use one of my custom SFZ files (you can see an example [here](https://github.com/lminiero/musescore-vpo-midi-actions/blob/master/VPO/Strings/0LM-1st-violin-SEC-KS-CC14.sfz)), which use CC14 as a trigger instead of the keyswitch in Virtual Playing Orchestra, I get the same, so CC14 is not listed as an option (unless `Unnamed` there will indeed "control" CC14?). It's not clear where the `label_cc` should be used, considering the same controller can activate different groups: what's the proper place for it?

Thanks again for the quick fix!

**Comment by paulfd** [comment]
It's switched using `locc` and `hicc`? I don't take these into account in the current patch; I stupidly did not think about keyswitch automation, my bad.

**Comment by paulfd** [comment]
You can check 967c55a614a61efd257ef834ec8b99eb77da3b1f

**Comment by lminiero** [comment]
That didn't work for me, back to the empty list :slightly_frowning_face:  I'll try again later today to see if I'm doing something wrong (I manually applied your last commit on the pull request I pulled before).

**Comment by paulfd** [comment]
I will install Ardour5, can you share the sfz file in question? I have the VPO samples lying around.

Jul 20, 2020 09:02:03 Lorenzo Miniero <notifications@github.com>:

**Comment by lminiero** [comment]
A simple example is this: https://github.com/lminiero/musescore-vpo-midi-actions/blob/master/VPO/Strings/0LM-1st-violin-SEC-KS-CC14.sfz. It's basically an editing of `1st-violin-SEC-PERF-KS-C2.sfz`, where I use MIDI CC 14 to act as a keyswitch instead of notes outside of the instrument boundaries like the `KS` scripts do.

**Comment by paulfd** [comment]
For me it indeed does not work, but on reloading the project the CCs do show properly. Can you confirm?

**Comment by paulfd** [comment]
OK after some discussion with Robin and Paul it seems that Ardour 5.12 assumes that the controls are fixed for the life of a plugin, so since we send an empty list at the beginning it stays empty until the plugin (instance I suppose?) is reloaded.

If you are willing to upgrade to Ardour 6---which brings numerous improvements as far as I understand!---it would be the best way as apparently it works well there and it is fully dynamic. If not, reloading the session after you set all the sfizz instances to their proper file is a workaround. I don't really see a cleaner way at the moment, sorry.

**Comment by AlexMitchellMus** [comment]
As this is an issue with Ardour 5.12, which is now EOL, and fixed in Ardour 6 I think this issue can be safely closed.

**Comment by lminiero** [comment]
Makes sense, and thanks for the clarification. I'm afraid my Fedora 31 is stuck to Ardour 5, so I'll have to wait until I upgrade my OS to see if Ardour 6 is available in Fedora 32/33. Thanks for your support!

**Comment by lminiero** [comment]
Apologies for bumping this old issue, but since I'm on vacation I started tinkering with this again :smile: 
I'm writing this here because I'm not sure if it's the same issue or a different one: I can open a new one if you prefer. I see there's a merged PR that references the issue I opened, but I already tried the latest version of the `develop` branch to no avail.

Since you mentioned `label_cc` in one of your previous posts, I managed to get Ardour to indeed list my controller properly, placing `label_cc` in a `<control>` section which is what the SFZ specification says:

```
<control>
label_cc14=CC based switch
```

That said, sending my CC messages through Ardour didn't have any effect, and I could actually not hear anything at all. As such, I decided to try and do some manual checks, using `sfizz_jack` to load one of my SFZ files from the command line, and then using [SendMIDI](https://github.com/gbevin/SendMIDI) (via `a2jmidi_bridge`, since SendMIDI only uses Alsa and sfizz only uses Jack) to manually send CC messages to sfizz, e.g.:

```
sendmidi dev playback cc 14 1
```

That didn't work either, while it worked when I opened the same SFZ file with LinuxSampler.

Then I noticed there was actually something weird in the logs when loading the SFZ file via `sfizz_jack` in the first place:

```
Total:
	Masters: 0
	Groups: 12
	Regions: 0
	Curves: 7
	PreloadedSamples: 0
```

Apparently, sfizz isn't recognizing any of the regions or samples, which probably explains why I can't hear anything. If I open one of the existing KS files from VPO (which I used as a basis for mine), e.g., `1st-violin-SEC-KS-C2.sfz`, the output is quite different:

```
Total:
	Masters: 0
	Groups: 12
	Regions: 171
	Curves: 7
	PreloadedSamples: 87
```

What could cause `sfizz` to miss/skip some of the `Regions` or `PreloadedSamples` in my version? It definitely worked some months ago (I created the repo with my SFZ files in March), so I guess something must have changed recently to cause this issue. I'm not familiar with the code internals so I wouldn't know what to look for myself, but I can try doing a `git bisect` to identify a commit that may have caused it to break, if you think it can help (I'll have to find one of the older commits that works first, of course). FYI, you can also test this with one of the files I mentioned in my previous posts here, like [0LM-1st-violin-SEC-KS-CC14.sfz](https://github.com/lminiero/musescore-vpo-midi-actions/blob/master/VPO/Strings/0LM-1st-violin-SEC-KS-CC14.sfz), that just needs to be put in the `Strings` folder of VPO along the others.

Thanks!

**Comment by jpcima** [comment]
The regions are absent when the sample files are not found.

I am copying your file into VPO and opening this with the jack client.
`./build/clients/sfizz_jack /opt/Virtual-Playing-Orchestra3/Strings/0LM-1st-violin-SEC-KS-CC14.sfz`

There are regions loaded, the file plays.
```
Total:
	Masters: 0
	Groups: 12
	Regions: 171
	Curves: 7
	PreloadedSamples: 87
```

Using LV2, anonymous controllers are recognized and listed.
```
		<ControlNameList Name="Controls">
			<Control Type="7bit" Number="1" Name="Unnamed CC 1" />
			<Control Type="7bit" Number="11" Name="Unnamed CC 11" />
			<Control Type="7bit" Number="14" Name="Unnamed CC 14" />
			<Control Type="7bit" Number="67" Name="Unnamed CC 67" />
		</ControlNameList>
```

**Comment by lminiero** [comment]
Thanks for the super-fast answer!

I have the problematic SFZ file in the same folder as the others, and the paths to the samples in them have not been changed (I verified with `diff`), so the cause seems to be different for me. I went back to commit 633159c5e1ed28c49e1e5573156f0ebae36c4265 (the one preceding #78 that introduced MIDNAM support) and the same SFZ file works on that old version of sfizz for me (so paths to sample files are indeed correct, or that wouldn't work either), but so did the commit that actually merged that PR (so it's not the MIDNAM support that broke it).

It's weird it works for you, as it definitely doesn't for me. I updated all the recursive submodules as well just to be sure, but nothing changed. Now that I know of some commits that do work, though, I can try a `git bisect`: it will be a slow process (I have to rebuild sfizz at every commit) but it might help identify what's causing the issue in my case.

**Comment by jpcima** [comment]
If you have ability to debug into this, I point you to where the region removal is.
(at various calls of `removeCurrentRegion`)

First one at 473, next at lines 479 and 511
https://github.com/sfztools/sfizz/blob/aba8d966ab5ae74dc51e56c14975a37cfc90f229/src/sfizz/Synth.cpp#L473

If you can add some print statements at these various points, this can help to identify the cause of problems.

**Comment by lminiero** [comment]
Thanks, I'll do that! As a developer myself (even if in a different space), I'm a big supporter of print-based debugging :smile: 
Now that I know where to dig, I'll hopefully manage to find some more useful info for you.

**Comment by lminiero** [comment]
Ok, I think I found out the cause of the issue. All the samples were failing in `resources.filePool.checkSampleId(region->sampleId)`, which looking at the code did indeed point to your note on the file being unavailable. Then I noticed that I actually had symbolic links for my own SFZ files, rather than physical copies: copying one of the files manually, did indeed get the expected output!

Not sure why it worked when SFZ files were symbolic links before, and not now, but at least I know the cause and can work around it. Thanks for the support and the great work!
